### PR TITLE
Alert on certain host status transitions

### PR DESCRIPTION
**Pull request by Slair1**
## Description

Add ability to create alert for host status transitions, such as from Alert->Up or Up->Disconnected, etc.  Right now some transitions may alert, but we wanted to know about any major host status transition.  

We also created a ConfigKey (alert.on.host.transiton) that defaults to the current behavior of not alerting on transitions.  But, by changing the setting to True, we are alerted if there are host transitions.  Which we find helpful - especially as once in a great while we'll have a host transition out of an Up status to something else.  But we never get an alert that it transitioned back to Up (even if it happened 1 second later)....

The ConfigKey works in conjunction with a flag set on the FSM's transitions.  If the TransitionAlertFlag for a particular transition is set to True AND the ConfigKey is set to True then the alert will be raised.

Let me know if 4.13 is OK or if you prefer a different branch instead of 4.13.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):
Our environment has it set to True, but it defaults to False.
image
image

## How Has This Been Tested?

Tested putting hosts through various states.  Made sure we received the emails and the Alerts appear in the UI.

### CODE REVIEW

File: api/src/main/java/com/cloud/host/Status.java [hunk 1]
```diff
@@ -119,45 +122,45 @@ public Status getNextStatus(Event e) throws NoTransitionException {
     static {
```
Comment (Slair1) [inline api/src/main/java/com/cloud/host/Status.java]: if a true flag is passed at the end, that is the transitionAlertFlag.  If it is True then an alert should be raised.  If it is False or not passed at all, then an alert should NOT be raised.

File: engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java [hunk 2]
```diff
@@ -1432,14 +1436,29 @@ public boolean agentStatusTransitTo(final HostVO host, final Status.Event e, fin
...
+                if (status_logger.isDebugEnabled()) {
+                    status_logger.debug("agentStatusTransitTo: old status - " + hostStatus + ", event - " + e + ", new status - " + nextStatus + " (" + (didTransit ? "success)" : "failed)"));
+                }
+                if (AlertOnHostTransitions.value() && hostStatus.getTransitionAlertFlag(e)) {
+                    final DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd H:m:s,SSS");
+                    final String hostShortDesc = host.getName() + " (id:" + host.getId() + ")";
+                    _alertMgr.sendAlert(AlertManager.AlertType.ALERT_TYPE_HOST, host.getDataCenterId(), host.getPodId(), host.getName() + " transit from " + hostStatus + " to " + nextStatus +  " (" + (didTransit ? "success)" : "failed)"),
+                        (dateFormatter.format(LocalDateTime.now())) + "\r\nTransit agent status with event " + e + " for host " + hostShortDesc + " from " + hostStatus + " to " + nextStatus + (didTransit ? " was successful" : " failed"));
+                }
+
```
Comment (Slair1) [inline engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java]: Above code is what raises the alert if both the ConfigKey is set and the TransitionAlertFlag is set for the particular Transition

File: utils/src/main/java/com/cloud/utils/fsm/StateMachine2.java [hunk 3]
```diff
@@ -50,7 +50,11 @@ public void addInitialTransition(E event, S toState) {
     }
 
     public void addTransition(S currentState, E event, S toState) {
-      addTransition(new Transition<S, E>(currentState, event, toState, null));
+      addTransition(currentState, event, toState, false);
+    }
+
+    public void addTransition(S currentState, E event, S toState, boolean alertFlag) {
+      addTransition(new Transition<S, E>(currentState, event, toState, null, alertFlag));
     }
```
Comment (Slair1) [inline utils/src/main/java/com/cloud/utils/fsm/StateMachine2.java]: Old constructor just calls new constructor with False for the alertFlag.  So we don't break any functionality using the previous constructor

File: utils/src/main/java/com/cloud/utils/fsm/StateMachine2.java [hunk 4]
```diff
@@ -171,15 +179,22 @@ public String toString() {
...
       public static enum Impact {
         USAGE
       }
 
       public Transition(S currentState, E event, S toState, List<Impact> impacts) {
+        this(currentState, event, toState, impacts, false);
+      }
+
+      public Transition(S currentState, E event, S toState, List<Impact> impacts, boolean alertFlag) {
         this.currentState = currentState;
```
Comment (Slair1) [inline utils/src/main/java/com/cloud/utils/fsm/StateMachine2.java]: Old constructor just calls new constructor with False for the alertFlag. So we don't break any functionality using the previous constructor

**Comment by DaanHoogland** [comment]
@Slair1 you marked this as new feature/enhancement. These only go on new releases. can you rebase to master?

**Comment by Slair1** [comment]
> @Slair1 you marked this as new feature/enhancement. These only go on new releases. can you rebase to master?

yep sure can.  we are also getting alerts from this feature for our secondary storage VM in one of our environments.  We are tracking down if it is a problem with this PR or if this PR just revealed an issue in that environment that we were previously unaware of.

**Pull request closed without merging**